pulsar-broker/src/main/java/org/apache/pulsar/common/naming/FlowOrQpsEquallyDivideBundleSplitAlgorithm.java

```diff
@@ -0,0 +1,129 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.common.naming;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import org.apache.pulsar.broker.namespace.NamespaceService;
+import org.apache.pulsar.common.policies.data.stats.TopicStatsImpl;
+
+
+/**
+ * Split algorithm based on flow or qps.
+ */
+public class FlowOrQpsEquallyDivideBundleSplitAlgorithm implements NamespaceBundleSplitAlgorithm {
+    private static final long MBytes = 1024 * 1024;
+
+    class TopicInfo {
+        String topicName;
+        double msgRate;
+        double throughput;
+
+        public TopicInfo(String topicName, double msgRate, double throughput) {
+            this.topicName = topicName;
+            this.msgRate = msgRate;
+            this.throughput = throughput;
+        }
+    }
+
+    @Override
+    public CompletableFuture<List<Long>> getSplitBoundary(BundleSplitOption bundleSplitOptionTmp) {
+        FlowOrQpsEquallyDivideBundleSplitOption bundleSplitOption =
+                (FlowOrQpsEquallyDivideBundleSplitOption) bundleSplitOptionTmp;
+        NamespaceService service = bundleSplitOption.getService();
+        NamespaceBundle bundle = bundleSplitOption.getBundle();
+        Map<String, TopicStatsImpl> topicStatsMap = bundleSplitOption.getTopicStatsMap();
+        int loadBalancerNamespaceBundleMaxMsgRate = bundleSplitOption.getLoadBalancerNamespaceBundleMaxMsgRate();
+        double diffThreshold = bundleSplitOption.getFlowOrQpsDifferenceThresholdPercentage() / 100.0;
+        long loadBalancerNamespaceBundleMaxBandwidthBytes = bundleSplitOption
+                .getLoadBalancerNamespaceBundleMaxBandwidthMbytes() * MBytes;
+
+
+        return service.getOwnedTopicListForNamespaceBundle(bundle).thenCompose(topics -> {
+            if (topics == null || topics.size() <= 1) {
+                return CompletableFuture.completedFuture(null);
+            }
+
+            double bundleThroughput = 0;
+            double bundleMsgRate = 0;
+            Map<Long, TopicInfo> topicInfoMap = new HashMap<>();
+            List<Long> topicHashList = new ArrayList<>(topics.size());
+            for (String topic : topics) {
+                TopicStatsImpl topicStats = topicStatsMap.get(topic);
+                if (topicStats == null) {
+                    continue;
+                }
+                double msgRateIn = topicStats.getMsgRateIn();
+                double msgRateOut = topicStats.getMsgRateOut();
```

Comment:
I think this one is based on the message, not the entry right?

Reply:
https://lists.apache.org/thread/df5v1q1xw1lo1y6n2wf3hp5zjoky0vmz The related discussion

Reply:
> I think this one is based on the message, not the entry right?

Yes, this is based on message

Reply:
> I think this one is based on the message, not the entry right?

Yes, this is based on the message, not the entry right.

Reply:
I'm not sure, is there currently a statistic entry rate

Reply:
I tried to calculate EntryRate with org.apache.pulsar.common.policies.data.stats.SubscriptionStatsImpl#chunkedMessageRate and org.apache.pulsar.common.policies.data.stats.PublisherStatsImpl#chunkedMessageRate.

@codelipenghui

Reply:
In org.apache.pulsar.broker.loadbalance.impl.ModularLoadManagerImpl#checkNamespaceBundleSplit, when selecting the bundle to be split, there is also a MessageRate condition, which is also the message rate used, not the entry rate:

<img width="1394" alt="image" src="https://user-images.githubusercontent.com/19296967/188468870-011b2b08-dd8d-4313-b784-ca54512381c1.png">

@codelipenghui

Reply:
Ok, I see. It could be a part to improve in a separate PR or PIP. Not a blocker for this PR

Reply:
OK，I will submit a new PIP to replace message rate with entry rate！